File: 554-brick-wall/ken/554.py
class Solution(object):
    def leastBricks(self, wall):
        """
        :type wall: List[List[int]]
        :rtype: int
        """

        n = len(wall)
        cnt = {}

        for line in wall:
            brickIdx = 0
            for i in xrange(len(line) - 1):
                brickIdx += line[i]
                line[i] = brickIdx

                if brickIdx not in cnt:
                    cnt[brickIdx] = 1
                else:
                    cnt[brickIdx] += 1

        if len(cnt) == 0:
            return n

        numLines = max(cnt.values())

        return n - numLines

a = Solution()
print a.leastBricks([[1,2,2,1],
 [3,1,2],
 [1,3,2],
 [2,4],
 [3,1,2],
 [1,3,1,1]])

print a.leastBricks([[1],[1],[1]])
print a.leastBricks([[1,1],[2],[1,1]])
print a.leastBricks([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1], [1, 1, 1, 1, 1, 1, 1, 1, 1, 1], [1, 1, 1, 1, 1, 1, 1, 1, 1, 1]])

File: 554-brick-wall/leochang/554.cpp
class Solution {
public:
	int leastBricks(vector<vector<int>>& wall) {
		if (0 == wall.size()) return 0;

		int maxWidth = 0;
		unordered_map<int, int> edges;

		for (auto row : wall) {
			int sum = 0;
			for (int i = 0; i < row.size(); i++) {
				sum += row[i];
				edges[sum]++;
			}
			maxWidth = sum;
		}

		int maxHit = 0;
		edges.erase(maxWidth);
		for (auto &edge : edges) {
			maxHit = max(maxHit, edge.second);
		}

		return wall.size() - maxHit;
	}
};
